diff --git a/CMakeLists.txt b/CMakeLists.txt
index 11565a1..c74fdf6 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -124,7 +124,8 @@ list(REMOVE_ITEM CEF_CXX_COMPILER_FLAGS -fno-rtti)
 list(REMOVE_ITEM CEF_CXX_COMPILER_FLAGS -std=c++17)
 list(REMOVE_ITEM CEF_COMPILER_FLAGS -fvisibility=hidden)
 list(REMOVE_ITEM CEF_COMPILER_FLAGS -fvisibility-inlines-hidden)
-list(APPEND CEF_CXX_COMPILER_FLAGS -std=c++20)
+list(REMOVE_ITEM CEF_CXX_COMPILER_FLAGS -Wno-undefined-var-template)
+list(APPEND CEF_CXX_COMPILER_FLAGS -std=c++2a)
 list(APPEND CEF_EXE_LINKER_FLAGS -rdynamic)
 # remove the vulkan library from the files to copy so it doesn't fail after removing it
 list(REMOVE_ITEM CEF_BINARY_FILES libvulkan.so.1)
diff --git a/CMakeModules/FindFFMPEG.cmake b/CMakeModules/FindFFMPEG.cmake
index 4ad1c79..313cebf 100644
--- a/CMakeModules/FindFFMPEG.cmake
+++ b/CMakeModules/FindFFMPEG.cmake
@@ -20,13 +20,6 @@
 
 include(FindPackageHandleStandardArgs)
 
-    find_package_handle_standard_args(FFMPEG
-        FOUND_VAR FFMPEG_FOUND
-        VERSION_VAR FFMPEG_VERSION
-        HANDLE_COMPONENTS
-        REQUIRED_VARS
-    )
-
 if(FFMPEG_LIBRARIES AND FFMPEG_INCLUDE_DIR)
   # in cache already
   set(FFMPEG_FOUND TRUE)
diff --git a/src/WallpaperEngine/Application/WallpaperApplication.cpp b/src/WallpaperEngine/Application/WallpaperApplication.cpp
index c9107af..6656e4f 100644
--- a/src/WallpaperEngine/Application/WallpaperApplication.cpp
+++ b/src/WallpaperEngine/Application/WallpaperApplication.cpp
@@ -688,8 +688,8 @@ void WallpaperApplication::setupOutput () {
 
 void WallpaperApplication::setupAudio () {
     // ensure audioprocessing is required by any background, and we have it enabled
-    const bool audioProcessingRequired = std::ranges::any_of (
-	this->m_backgrounds, [] (const std::pair<const std::string, ProjectUniquePtr>& pair) -> bool {
+    const bool audioProcessingRequired = std::any_of (
+	this->m_backgrounds.begin (), this->m_backgrounds.end (), [] (const std::pair<const std::string, ProjectUniquePtr>& pair) -> bool {
 	    return pair.second->supportsAudioProcessing;
 	}
     );
diff --git a/src/WallpaperEngine/Audio/Drivers/SDLAudioDriver.cpp b/src/WallpaperEngine/Audio/Drivers/SDLAudioDriver.cpp
index 297a606..4871341 100644
--- a/src/WallpaperEngine/Audio/Drivers/SDLAudioDriver.cpp
+++ b/src/WallpaperEngine/Audio/Drivers/SDLAudioDriver.cpp
@@ -19,7 +19,7 @@ void audio_callback (void* userdata, uint8_t* streamData, int length) {
 
     SDL_LockMutex (driver->getStreamMutex ());
 
-    for (const auto& buffer : driver->getStreams () | std::views::values) {
+    for (const auto& [key, buffer] : driver->getStreams ()) {
 	uint8_t* streamDataPointer = streamData;
 	int streamLength = length;
 
diff --git a/src/WallpaperEngine/Data/Builders/ColorBuilder.cpp b/src/WallpaperEngine/Data/Builders/ColorBuilder.cpp
index e13a2d5..ac00b33 100644
--- a/src/WallpaperEngine/Data/Builders/ColorBuilder.cpp
+++ b/src/WallpaperEngine/Data/Builders/ColorBuilder.cpp
@@ -2,7 +2,7 @@
 #include "VectorBuilder.h"
 
 #include <algorithm>
-#include <format>
+#include <sstream>
 #include <glm/vec3.hpp>
 
 const WallpaperEngine::Data::Model::Color WallpaperEngine::Data::Builders::ColorBuilder::White
@@ -17,7 +17,7 @@ WallpaperEngine::Data::Builders::ColorBuilder::parse (const std::string& value,
     // replace the actual separators with spaces to normalize them
     if (copy.find (',') != std::string::npos) {
 	// replace comma separator with spaces so it's
-	std::ranges::replace (copy, ',', ' ');
+	std::replace (copy.begin (), copy.end (), ',', ' ');
     }
 
     // hex colors should be converted to int colors
@@ -27,15 +27,16 @@ WallpaperEngine::Data::Builders::ColorBuilder::parse (const std::string& value,
 	// expand short css notation into the right one
 	// support for css notation
 	if (number.size () == 3) {
-	    number = std::format (
-		"{}{}{}{}{}{}{:02x}", number.at (0), number.at (0), number.at (1), number.at (1), number.at (2),
-		number.at (2), static_cast<int> (alpha * 255)
-	    );
+	    std::stringstream ss;
+	    ss << number.at (0) << number.at (0) << number.at (1) << number.at (1) << number.at (2) << number.at (2);
+	    char alphaHex[10];
+	    snprintf (alphaHex, sizeof (alphaHex), "%02x", static_cast<int> (alpha * 255));
+	    ss << alphaHex;
+	    number = ss.str ();
 	} else if (number.size () == 4) {
-	    number = std::format (
-		"{}{}{}{}{}{}{}{}", number.at (0), number.at (0), number.at (1), number.at (1), number.at (2),
-		number.at (2), number.at (3), number.at (3)
-	    );
+	    std::stringstream ss;
+	    ss << number.at (0) << number.at (0) << number.at (1) << number.at (1) << number.at (2) << number.at (2) << number.at (3) << number.at (3);
+	    number = ss.str ();
 	} else if (number.size () != 6 && number.size () != 8) {
 	    sLog.exception ("Invalid CSS color notation for ", value);
 	}
diff --git a/src/WallpaperEngine/Data/Parsers/ProjectParser.cpp b/src/WallpaperEngine/Data/Parsers/ProjectParser.cpp
index 02de380..8896a20 100644
--- a/src/WallpaperEngine/Data/Parsers/ProjectParser.cpp
+++ b/src/WallpaperEngine/Data/Parsers/ProjectParser.cpp
@@ -30,7 +30,7 @@ ProjectUniquePtr ProjectParser::parse (const JSON& data, AssetLocatorUniquePtr c
     }
 
     // lowercase for consistency
-    std::ranges::transform (type, type.begin (), tolower);
+    std::transform (type.begin (), type.end (), type.begin (), tolower);
 
     auto result = std::make_unique<Project> (Project {
 	.title = data.require<std::string> ("title", "Project title missing"),
diff --git a/src/WallpaperEngine/FileSystem/Adapters/Package.cpp b/src/WallpaperEngine/FileSystem/Adapters/Package.cpp
index fb34be6..2ca17e9 100644
--- a/src/WallpaperEngine/FileSystem/Adapters/Package.cpp
+++ b/src/WallpaperEngine/FileSystem/Adapters/Package.cpp
@@ -15,7 +15,7 @@ using namespace WallpaperEngine::FileSystem::Adapters;
 
 ReadStreamSharedPtr PackageAdapter::open (const std::filesystem::path& path) const {
     // find the file entry
-    const auto it = std::ranges::find_if (this->package->files, [&path] (const auto& file) {
+    const auto it = std::find_if (this->package->files.begin (), this->package->files.end (), [&path] (const auto& file) {
 	return file->filename == path.string ();
     });
 
diff --git a/src/WallpaperEngine/Media/MediaSource.cpp b/src/WallpaperEngine/Media/MediaSource.cpp
index f3c6126..a96b6ce 100644
--- a/src/WallpaperEngine/Media/MediaSource.cpp
+++ b/src/WallpaperEngine/Media/MediaSource.cpp
@@ -1,6 +1,6 @@
 #include "MediaSource.h"
 
-#include <ranges>
+#include <memory>
 
 using namespace WallpaperEngine::Media;
 
@@ -68,13 +68,13 @@ std::function<void ()> MediaSource::addAlbumArtListener (std::function<void (con
 }
 
 void MediaSource::fireMetadataListeners () {
-    for (auto& listener : m_metadataListeners | std::views::values) {
+    for (auto& [key, listener] : m_metadataListeners) {
 	listener (m_mediaInfo);
     }
 }
 
 void MediaSource::fireAlbumArtListeners () {
-    for (auto& listener : m_albumArtListeners | std::views::values) {
+    for (auto& [key, listener] : m_albumArtListeners) {
 	listener (m_mediaInfo);
     }
 }
\ No newline at end of file
diff --git a/src/WallpaperEngine/Media/MediaSource.h b/src/WallpaperEngine/Media/MediaSource.h
index 300a30c..19663c6 100644
--- a/src/WallpaperEngine/Media/MediaSource.h
+++ b/src/WallpaperEngine/Media/MediaSource.h
@@ -3,6 +3,7 @@
 #include <chrono>
 #include <functional>
 #include <map>
+#include <memory>
 #include <string>
 
 namespace WallpaperEngine::Media {
diff --git a/src/WallpaperEngine/Render/AlbumTexture.cpp b/src/WallpaperEngine/Render/AlbumTexture.cpp
index 2769c0b..dfbafa2 100644
--- a/src/WallpaperEngine/Render/AlbumTexture.cpp
+++ b/src/WallpaperEngine/Render/AlbumTexture.cpp
@@ -96,7 +96,7 @@ void AlbumTexture::load () const {
     this->m_width = 0;
     this->m_height = 0;
 
-    for (const auto& project : this->getContext ().getApp ().getBackgrounds () | std::views::values) {
+    for (const auto& [key, project] : this->getContext ().getApp ().getBackgrounds ()) {
 	try {
 	    // try to open the file in any of the asset locators
 	    auto contents = project->assetLocator->read ("$mediaThumbnail");
diff --git a/src/WallpaperEngine/Render/Drivers/VideoFactories.cpp b/src/WallpaperEngine/Render/Drivers/VideoFactories.cpp
index a7045cb..0ea3240 100644
--- a/src/WallpaperEngine/Render/Drivers/VideoFactories.cpp
+++ b/src/WallpaperEngine/Render/Drivers/VideoFactories.cpp
@@ -41,13 +41,14 @@ void VideoFactories::registerFullscreenDetector (
 std::vector<std::string> VideoFactories::getRegisteredDrivers () const {
     std::vector<std::string> result;
 
-    for (const auto& sessionTypeToFactory : this->m_driverFactories | std::views::values) {
-	for (const auto& xdgSessionType : sessionTypeToFactory | std::views::keys) {
+    for (const auto& [sessionType, sessionTypeToFactory] : this->m_driverFactories) {
+	for (const auto& [xdgSessionType, factory] : sessionTypeToFactory) {
 	    result.push_back (xdgSessionType);
 	}
     }
 
-    const auto last = std::ranges::unique (result).begin ();
+    std::sort (result.begin (), result.end ());
+    const auto last = std::unique (result.begin (), result.end ());
 
     result.erase (last, result.end ());
 
diff --git a/src/WallpaperEngine/Render/Drivers/WaylandOpenGLDriver.cpp b/src/WallpaperEngine/Render/Drivers/WaylandOpenGLDriver.cpp
index e4b722d..de49c67 100644
--- a/src/WallpaperEngine/Render/Drivers/WaylandOpenGLDriver.cpp
+++ b/src/WallpaperEngine/Render/Drivers/WaylandOpenGLDriver.cpp
@@ -342,7 +342,7 @@ void WaylandOpenGLDriver::setupOutputLayerSurfaces () {
 
 	sLog.error ("Requested: ");
 
-	for (const auto& o : m_context.settings.general.screenBackgrounds | std::views::keys) {
+	for (const auto& [o, value] : m_context.settings.general.screenBackgrounds) {
 	    sLog.error ("  ", o);
 	}
 
@@ -400,7 +400,7 @@ void WaylandOpenGLDriver::dispatchEventQueue () {
     if (!initialized) {
 	initialized = true;
 
-	for (const auto& viewport : this->getOutput ().getViewports () | std::views::values) {
+	for (const auto& [key, viewport] : this->getOutput ().getViewports ()) {
 	    this->getApp ().update (viewport);
 	}
     }
diff --git a/src/WallpaperEngine/Render/Objects/Effects/CPass.cpp b/src/WallpaperEngine/Render/Objects/Effects/CPass.cpp
index e93c293..38b30fe 100644
--- a/src/WallpaperEngine/Render/Objects/Effects/CPass.cpp
+++ b/src/WallpaperEngine/Render/Objects/Effects/CPass.cpp
@@ -327,7 +327,7 @@ void CPass::bindTextureOverrides (uint32_t currentTexture, std::shared_ptr<const
 
 void CPass::setupRenderReferenceUniforms () {
     // add reference uniforms
-    for (const auto& value : this->m_referenceUniforms | std::views::values) {
+    for (const auto& [key, value] : this->m_referenceUniforms) {
 	switch (value->type) {
 	    case Double:
 		glUniform1d (value->id, *static_cast<const double*> (*value->value));
@@ -363,7 +363,7 @@ void CPass::setupRenderReferenceUniforms () {
 
 void CPass::setupRenderUniforms () {
     // add uniforms
-    for (const auto& value : this->m_uniforms | std::views::values) {
+    for (const auto& [key, value] : this->m_uniforms) {
 	switch (value->type) {
 	    case Double:
 		glUniform1dv (value->id, value->count, static_cast<const double*> (value->value));
@@ -445,7 +445,7 @@ void CPass::cleanupRenderSetup () {
     glBindTexture (GL_TEXTURE_2D, 0);
 
     // continue on the map from the second texture
-    for (const auto& index : this->m_textures | std::views::keys) {
+    for (const auto& [index, texture] : this->m_textures) {
 	glActiveTexture (GL_TEXTURE0 + index);
 	glBindTexture (GL_TEXTURE_2D, 0);
     }
diff --git a/src/WallpaperEngine/Render/RenderContext.cpp b/src/WallpaperEngine/Render/RenderContext.cpp
index af0112e..9c89665 100644
--- a/src/WallpaperEngine/Render/RenderContext.cpp
+++ b/src/WallpaperEngine/Render/RenderContext.cpp
@@ -45,7 +45,7 @@ void RenderContext::setWallpaper (const std::string& display, std::shared_ptr<CW
 }
 
 void RenderContext::setPause (const bool newState) const {
-    for (const auto& wallpaper : this->m_wallpapers | std::views::values) {
+    for (const auto& [key, wallpaper] : this->m_wallpapers) {
 	wallpaper->setPause (newState);
     }
 }
diff --git a/src/WallpaperEngine/Render/Shaders/ShaderUnit.cpp b/src/WallpaperEngine/Render/Shaders/ShaderUnit.cpp
index fcc03fc..c4cf1dd 100644
--- a/src/WallpaperEngine/Render/Shaders/ShaderUnit.cpp
+++ b/src/WallpaperEngine/Render/Shaders/ShaderUnit.cpp
@@ -662,7 +662,7 @@ const std::string& ShaderUnit::compile () {
 
     for (const auto& [name, value] : this->m_overrideCombos) {
 	std::string uppercase;
-	std::ranges::transform (name, std::back_inserter (uppercase), ::toupper);
+	std::transform (name.begin (), name.end (), std::back_inserter (uppercase), ::toupper);
 
 	if (!addedCombos.contains (uppercase)) {
 	    this->m_final += DEFINE_COMBO (uppercase, value);
@@ -673,7 +673,7 @@ const std::string& ShaderUnit::compile () {
     // now add all the combos to the source
     for (const auto& [name, value] : this->m_combos) {
 	std::string uppercase;
-	std::ranges::transform (name, std::back_inserter (uppercase), ::toupper);
+	std::transform (name.begin (), name.end (), std::back_inserter (uppercase), ::toupper);
 
 	if (!addedCombos.contains (uppercase)) {
 	    this->m_final += DEFINE_COMBO (uppercase, value);
@@ -683,7 +683,7 @@ const std::string& ShaderUnit::compile () {
 
     for (const auto& [name, value] : this->m_discoveredCombos) {
 	std::string uppercase;
-	std::ranges::transform (name, std::back_inserter (uppercase), ::toupper);
+	std::transform (name.begin (), name.end (), std::back_inserter (uppercase), ::toupper);
 
 	if (!addedCombos.contains (uppercase)) {
 	    this->m_final += DEFINE_COMBO (uppercase, value);
@@ -694,7 +694,7 @@ const std::string& ShaderUnit::compile () {
     if (this->m_link != nullptr) {
 	for (const auto& [name, value] : this->m_link->getCombos ()) {
 	    std::string uppercase;
-	    std::ranges::transform (name, std::back_inserter (uppercase), ::toupper);
+	    std::transform (name.begin (), name.end (), std::back_inserter (uppercase), ::toupper);
 
 	    if (!addedCombos.contains (uppercase)) {
 		this->m_final += DEFINE_COMBO (uppercase, value);
@@ -704,7 +704,7 @@ const std::string& ShaderUnit::compile () {
 
 	for (const auto& [name, value] : this->m_link->getDiscoveredCombos ()) {
 	    std::string uppercase;
-	    std::ranges::transform (name, std::back_inserter (uppercase), ::toupper);
+	    std::transform (name.begin (), name.end (), std::back_inserter (uppercase), ::toupper);
 
 	    if (!addedCombos.contains (uppercase)) {
 		this->m_final += DEFINE_COMBO (uppercase, value);
diff --git a/src/WallpaperEngine/Render/TextureCache.cpp b/src/WallpaperEngine/Render/TextureCache.cpp
index 862c5f0..d25451f 100644
--- a/src/WallpaperEngine/Render/TextureCache.cpp
+++ b/src/WallpaperEngine/Render/TextureCache.cpp
@@ -62,7 +62,7 @@ std::shared_ptr<const TextureProvider> TextureCache::resolve (const std::string&
     }
 
     // search for the texture in all the different containers just in case
-    for (const auto& project : this->getContext ().getApp ().getBackgrounds () | std::views::values) {
+    for (const auto& [key, project] : this->getContext ().getApp ().getBackgrounds ()) {
 	try {
 	    const auto contents = project->assetLocator->texture (filename);
 	    auto stream = BinaryReader (contents);
diff --git a/src/WallpaperEngine/Scripting/EngineObject.cpp b/src/WallpaperEngine/Scripting/EngineObject.cpp
index b8399fa..3549efd 100644
--- a/src/WallpaperEngine/Scripting/EngineObject.cpp
+++ b/src/WallpaperEngine/Scripting/EngineObject.cpp
@@ -2,8 +2,6 @@
 #include "ScriptEngine.h"
 #include "WallpaperEngine/Logging/Log.h"
 
-#include <ranges>
-
 using namespace WallpaperEngine::Scripting;
 
 extern float g_Time;
@@ -260,7 +258,7 @@ void EngineObject::tick () {
     const auto now = std::chrono::steady_clock::now ();
 
     // check any interval and run them if needed
-    for (auto& timeout : this->m_intervals | std::views::values) {
+    for (auto& [key, timeout] : this->m_intervals) {
 	if (timeout.next > now) {
 	    continue;
 	}
diff --git a/src/WallpaperEngine/Scripting/ScriptEngine.cpp b/src/WallpaperEngine/Scripting/ScriptEngine.cpp
index e690b92..3d48991 100644
--- a/src/WallpaperEngine/Scripting/ScriptEngine.cpp
+++ b/src/WallpaperEngine/Scripting/ScriptEngine.cpp
@@ -26,7 +26,6 @@
 #include <future>
 #include <optional>
 #include <poll.h>
-#include <ranges>
 #include <signal.h>
 #include <spawn.h>
 #include <sstream>
@@ -251,7 +250,7 @@ ScriptEngine::ScriptEngine (Wallpapers::CScene& scene, Media::MediaSource& media
 ScriptEngine::~ScriptEngine () {
     this->m_unregisterMediaUpdateCallback ();
 
-    for (const auto& module : this->m_scriptModules | std::views::values) {
+    for (const auto& [key, module] : this->m_scriptModules) {
 	JS_FreeValue (this->m_context, module.module);
     }
 
@@ -620,7 +619,7 @@ void ScriptEngine::tick () {
     // run any pending notifications
 
     // run all update methods
-    for (auto& module : this->m_scriptModules | std::views::values) {
+    for (auto& [key, module] : this->m_scriptModules) {
 	this->m_runningModule = &module;
 
 	JSValue args[] = { this->dynamicToJs (module.value) };
@@ -681,7 +680,7 @@ void ScriptEngine::notifyMediaUpdate (const Media::MediaSource::MediaInfo& media
     JSValue mediaTimelineArgs[] = { mediaTimelineEvent };
     JSValue mediaThumbnailArgs[] = { mediaThumbnailEvent };
 
-    for (auto& module : this->m_scriptModules | std::views::values) {
+    for (auto& [key, module] : this->m_scriptModules) {
 	// call all methods
 	JSValue result1 = this->call (module.module, 1, propertiesArgs, "mediaPropertiesChanged");
 	JSValue result2 = this->call (module.module, 1, playbackArgs, "mediaPlaybackChanged");
diff --git a/src/WallpaperEngine/Scripting/ScriptableObject.cpp b/src/WallpaperEngine/Scripting/ScriptableObject.cpp
index bfd4be8..e415396 100644
--- a/src/WallpaperEngine/Scripting/ScriptableObject.cpp
+++ b/src/WallpaperEngine/Scripting/ScriptableObject.cpp
@@ -3,8 +3,6 @@
 #include "ScriptEngine.h"
 #include "WallpaperEngine/Data/Utils/ScopeGuard.h"
 
-#include <ranges>
-
 using namespace WallpaperEngine::Render;
 using namespace WallpaperEngine::Scripting;
 
diff --git a/src/WallpaperEngine/WebBrowser/CEF/SubprocessApp.cpp b/src/WallpaperEngine/WebBrowser/CEF/SubprocessApp.cpp
index 7b10769..002f906 100644
--- a/src/WallpaperEngine/WebBrowser/CEF/SubprocessApp.cpp
+++ b/src/WallpaperEngine/WebBrowser/CEF/SubprocessApp.cpp
@@ -6,14 +6,14 @@ using namespace WallpaperEngine::WebBrowser::CEF;
 
 SubprocessApp::SubprocessApp (WallpaperEngine::Application::WallpaperApplication& application) :
     m_application (application) {
-    for (const auto& info : this->m_application.getBackgrounds () | std::views::values) {
+    for (const auto& [key, info] : this->m_application.getBackgrounds ()) {
 	this->m_handlerFactories[info->workshopId] = new WPSchemeHandlerFactory (*info);
     }
 }
 
 void SubprocessApp::OnRegisterCustomSchemes (CefRawPtr<CefSchemeRegistrar> registrar) {
     // register all the needed schemes, "wp" + the background id is going to be our scheme
-    for (const auto& workshopId : this->m_handlerFactories | std::views::keys) {
+    for (const auto& [workshopId, factory] : this->m_handlerFactories) {
 	registrar->AddCustomScheme (
 	    WPSchemeHandlerFactory::generateSchemeName (workshopId),
 	    CEF_SCHEME_OPTION_STANDARD | CEF_SCHEME_OPTION_SECURE | CEF_SCHEME_OPTION_FETCH_ENABLED
